2025/lab/Keys/Week7_Missingness_Robustness.Rmd

```diff
@@ -0,0 +1,551 @@
+
+Of course, these are not simple fixes. You run the risk of pushing too many things to the middle, especially when it isn't warranted or even useful. Make sure you know your data when you do this.
+
+The function is `naniar::impute_mean()`. If you think your data are skewed, you can also use `impute_median()` instead.
+
+```{r central-imp}
+gss <- gss %>% 
+  mutate(imp_income = impute_median(income),
+         imp_loginc = impute_mean(logincome))
+
+gss %>% select(income, imp_income, logincome, imp_loginc) %>%
+  datasummary_skim()
+```
+
+Notice how imputing using median and mean puts more density around these values and reduces the standard deviation. This is especially visible in the density plots below.
+
+```{r imp-density}
+incomeplot <- gss %>% 
+  ggplot() + 
+  geom_density(aes(x = income, color = "Original")) +
+  geom_density(aes(x = imp_income, color = "Imputed")) +
+  theme_minimal() + ggtitle("INCOME")
+logimcomeplot <- gss %>% 
+  ggplot() + 
+  geom_density(aes(x = logincome, color = "Original")) +
+  geom_density(aes(x = imp_loginc, color = "Imputed")) +
+  theme_minimal() + ggtitle("LOGINCOME")
+
+cowplot::plot_grid(incomeplot, logimcomeplot, nrow = 2)
+  
+```
+
+### Regression Imputation
+
+Regression imputation is like mean imputation, but using multiple variables. It runs a regression equation on each missing observation to estimate the value of the missing data. Now, we can account for differences in missingness based on relationships with other variables.
+
+(If you're wondering how R does this for categorical variables, we'll get to it after spring break.)
+
+### Multivariate Imputation by Chained Equations (MICE)
+
+Multiple imputation is a term that means we're doing (regression) imputation multiple times.
+
+The benefit is that we no longer have to settle for just one version of the final imputed variable. We can now make a series of datasets, all with imputed data on the missing values. MICE is just a case of multiple imputation that uses chained equations.
+
+In short: We mean impute values for all variables, use these to regression impute the missing values of them, then repeat until we no longer are using the mean imputed values. Of course, regression imputation gives different values than mean imputation, so we repeat the procedure over and over until the values "converge," another way of saying they stop changing.
+
+If you're curious, more detail on the steps that make these equations chained can be found here: <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3074241/#mpr329-sec-0002title>
+
+Below, we use the `mice` command to create five new, imputed datasets. (If we only wanted one dataset, a la normal regression imputation, we set `m = 1`.) The `maxit` option tells it how many iterations it should do before giving up. The default is 5, but I set it to 2 here for the same of speed. 
```

Comment:
@brenthoagland Is it really slow with 5 iterations? If this is a problem then 2 is fine, but I don't want to encourage students to reduce iterations for things like this. If it computes in reasonable time, say <1min, then remove this and just use the default.